driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java

```diff
@@ -19,203 +19,145 @@
         this.maxTimeMS = maxTimeMS;
         this.decoder = decoder;
         this.comment = comment;
-        this.maxWireVersion = connection.getDescription().getMaxWireVersion();
-        this.firstBatchEmpty = initialCommandCursorResult.getResults().isEmpty();
-
-        this.operationTime = initialCommandCursorResult.getOperationTime();
-        if (cursor.get() != null) {
-            this.connectionSource = notNull("connectionSource", connectionSource).retain();
-            assertNotNull(connection);
-            if (limitReached()) {
-                killCursor(connection);
-            } else {
-                if (connectionSource.getServerDescription().getType() == ServerType.LOAD_BALANCER) {
-                    this.pinnedConnection = connection.retain();
-                    this.pinnedConnection.markAsPinned(Connection.PinningMode.CURSOR);
-                }
-            }
-        }
-        logQueryResult(initialCommandCursorResult);
-    }
-
-    /**
-     * {@inheritDoc}
-     * <p>
-     * From the perspective of the code external to this class, this method is idempotent as required by its specification.
-     * However, if this method sets {@link #isClosePending},
-     * then it must be called by {@code this} again to release resources.
-     * This behavior does not violate externally observable idempotence because this method is allowed to release resources "eventually".
-     */
-    @Override
-    public void close() {
-        boolean doClose = withLock(lock, () -> {
-            if (isOperationInProgress) {
-                isClosePending = true;
-                return false;
-            } else if (!isClosed) {
-                isClosed = true;
-                isClosePending = false;
-                return true;
-            }
-            return false;
-        });
+        this.maxWireVersion = connectionDescription.getMaxWireVersion();
+        this.firstBatchEmpty = commandCursorResult.getResults().isEmpty();
 
-        if (doClose) {
-            killCursorOnClose();
-        }
+        AsyncConnection connectionToPin = connectionDescription.getServerType() == ServerType.LOAD_BALANCER ? connection : null;
+        resourceManager = new ResourceManager(namespace, connectionSource, connectionToPin, commandCursorResult.getServerCursor());
     }
 
     @Override
     public void next(final SingleResultCallback<List<T>> callback) {
         if (isClosed()) {
-            callback.onResult(null, new MongoException("next() called after the cursor was closed."));
-        } else if (initialCommandCursorResult != null && !firstBatchEmpty) {
-            // May be empty for a tailable cursor
-            List<T> results = initialCommandCursorResult.getResults();
-            initialCommandCursorResult = null;
-            if (getServerCursor() == null) {
-                close();
+            callback.onResult(null, new MongoException(MESSAGE_IF_CLOSED_AS_CURSOR));
+            return;
+        }
```

Comment:
`tryStartOperation` in `execute` already handles the situation when `AsyncCommandBatchCursor` is closed in a non-racy way. Let's remove this code.

Reply:
Ack